src/node/stage/approved.rs

```diff
@@ -300,115 +296,273 @@ impl Approved {
+            Variant::NodeApproval(_) | Variant::BootstrapResponse(_) | Variant::Ping => {
+                Ok(MessageStatus::Useless)
+            }
         }
     }
 
-    pub fn verify_message(&self, msg: &Message) -> Result<bool> {
-        self.verify_message_quiet(msg).map_err(|error| {
-            messages::log_verify_failure(msg, &error, self.shared_state.sections.keys());
-            error
-        })
+    // Ignore stale GenesisUpdates
+    fn should_handle_genesis_update(
+        &self,
+        our_id: &PublicId,
+        genesis_prefix_info: &GenesisPrefixInfo,
+    ) -> bool {
+        !self.is_our_elder(our_id)
+            && genesis_prefix_info.parsec_version > self.genesis_prefix_info.parsec_version
+    }
+
+    // Ignore `JoinRequest` if we are not elder unless the join request is outdated in which case we
+    // reply with `BootstrapResponse::Join` with the up-to-date info (see `handle_join_request`).
+    fn should_handle_join_request(&self, our_id: &PublicId, req: &JoinRequest) -> bool {
+        self.is_our_elder(our_id) || req.section_key != *self.shared_state.our_history.last_key()
+    }
+
+    // If elder, always handle UserMessage, otherwise handle it only if addressed directly to us
+    // as a node.
+    fn should_handle_user_message(&self, our_id: &PublicId, dst: &DstLocation) -> bool {
+        self.is_our_elder(our_id) || dst.as_node().ok() == Some(our_id.name())
+    }
+
+    fn verify_message(&self, msg: &Message) -> Result<bool> {
+        match msg.verify(self.shared_state.sections.keys()) {
+            Ok(VerifyStatus::Full) => Ok(true),
+            Ok(VerifyStatus::Unknown) => Ok(false),
+            Err(error) => {
+                messages::log_verify_failure(msg, &error, self.shared_state.sections.keys());
+                Err(error)
+            }
+        }
+    }
+
+    /// Handle message whose trust we can't establish because its proof contains only keys we don't
+    /// know.
+    pub fn handle_untrusted_message(
+        &self,
+        core: &mut Core,
+        sender: Option<SocketAddr>,
+        msg: Message,
+    ) -> Result<()> {
+        let src = msg.src.location();
+        let bounce_dst = src.to_dst();
+        let bounce_dst_key = self
+            .shared_state
+            .sections
+            .key_by_location(&bounce_dst)
+            .copied();
+
+        if bounce_dst_key.is_none() {
+            trace!("Untrusted message from unknown src {:?} - discarding", src);
+            return Ok(());
+        }
+
+        let bounce_msg = Message::single_src(
+            &core.full_id,
+            bounce_dst,
+            bounce_dst_key,
+            Variant::BouncedUntrustedMessage(Box::new(msg)),
+        )?;
+        let bounce_msg = bounce_msg.to_bytes()?;
+
+        if let Some(sender) = sender {
+            core.send_message_to_target(&sender, bounce_msg)
+        } else {
+            self.send_message_to_our_elders(core, bounce_msg)
```

Comment:
Not sure. `sender` is `None` only if this is accumulated message, so that would mean there would have to be a `quorum` of malicious nodes to do this. `quorum` is currently 1/3 but we also work with the assumption that **less** than 1/3 of the nodes are malicious, so under this assumption this is not possible. 

There is, however, one issue related to this. Currently when we accumulate the messages, we take the proof from the first `MessageSignature` we receive and ignore the others. So this means if a malicious node somehow get its `MessageSignature` to be received first, it can purposefully include a proof that is too short and trigger this issue. This however, seems like too much effort if all you want to do is to spam the section. I'll create an issue for it regardless.

Reply:
https://github.com/maidsafe/routing/issues/2108